Update fields in ocean monthly output

mark-petersen opened the pull request:
With V2 just around this corner, it is a good time to double-check our variables in the monthly time-averaged output stream.

If all fields in ocean monthly time-averaged file
`mpaso.hist.am.timeSeriesStatsMonthly.$Y-$M-$D.nc`
are compared, then this is non-bit-for-bit.

mark-petersen commented:
So far, I've only added zMid, as that addition was requested for V1 output as a post-processing step. 

Please look through the current output list and make suggestions to add or remove. For your convenience:

<details>
<summary> Monthly time average,  in: components/mpas-ocean/cime_config/buildnml
</summary> <p>

```
 776         lines.append('<stream name="timeSeriesStatsMonthlyOutput"')
 777         lines.append('        type="output"')
 778         if ocn_grid.startswith("oRRS1"):
 779             lines.append('        io_type="pnetcdf,cdf5"')
 780         else:
 781             lines.append('        io_type="{}"'.format(ocn_pio_typename))
 782
 783         lines.append('        filename_template="mpaso.hist.am.timeSeriesStatsMonthly.$Y-$M-$D.nc"')
 784         lines.append('        filename_interval="00-01-00_00:00:00"')
 785         lines.append('        reference_time="01-01-01_00:00:00"')
 786         lines.append('        output_interval="00-01-00_00:00:00"')
 787         lines.append('        clobber_mode="truncate"')
 788         lines.append('        packages="timeSeriesStatsMonthlyAMPKG"')
 789         lines.append('        runtime_format="single_file">')
 790         lines.append('')
 791         lines.append('    <var name="daysSinceStartOfSim"/>')
 792         lines.append('    <var name="binBoundaryMerHeatTrans"/>')
 793         lines.append('    <var name="ssh"/>')
 794         lines.append('    <var_struct name="tracers"/>')
 795         lines.append('    <var name="velocityMeridional"/>')
 796         lines.append('    <var name="velocityZonal"/>')
 797         lines.append('    <var name="layerThickness"/>')
 798         lines.append('    <var name="zMid"/>')
 799         lines.append('    <var name="density"/>')
 800         lines.append('    <var name="potentialDensity"/>')
 801         lines.append('    <var name="windStressZonal"/>')
 802         lines.append('    <var name="windStressMeridional"/>')
 803         lines.append('    <var name="frazilLayerThicknessTendency"/>')
 804         lines.append('    <var_array name="avgValueWithinOceanRegion"/>')
 805         lines.append('    <var_array name="avgValueWithinOceanLayerRegion"/>')
 806         lines.append('    <var_array name="avgValueWithinOceanVolumeRegion"/>')
 807         lines.append('    <var name="meridionalHeatTransportLatZ"/>')
 808         lines.append('    <var name="meridionalHeatTransportLat"/>')
 809         lines.append('    <var_array name="minGlobalStats"/>')
 810         lines.append('    <var_array name="maxGlobalStats"/>')
 811         lines.append('    <var_array name="sumGlobalStats"/>')
 812         lines.append('    <var_array name="rmsGlobalStats"/>')
 813         lines.append('    <var_array name="avgGlobalStats"/>')
 814         lines.append('    <var_array name="vertSumMinGlobalStats"/>')
 815         lines.append('    <var_array name="vertSumMaxGlobalStats"/>')
 816         lines.append('    <var name="tThreshMLD"/>')
 817         lines.append('    <var name="dThreshMLD"/>')
 818         lines.append('    <var name="normalVelocity"/>')
 819         lines.append('    <var name="vertVelocityTop"/>')
 820         if not ocn_grid.startswith("oRRS1"):
 821             lines.append('    <var name="normalGMBolusVelocity"/>')
 822             lines.append('    <var name="vertGMBolusVelocityTop"/>')
 823             lines.append('    <var name="GMBolusVelocityZonal"/>')
 824             lines.append('    <var name="GMBolusVelocityMeridional"/>')
 825             lines.append('    <var name="kappaGM3D"/>')
 826             lines.append('    <var name="cGMphaseSpeed"/>')
 827             lines.append('    <var name="velocityZonalTimesTemperature_GM"/>')
 828             lines.append('    <var name="velocityMeridionalTimesTemperature_GM"/>')
 829
 830         lines.append('    <var_struct name="tracersSurfaceFlux"/>')
 831         lines.append('    <var name="penetrativeTemperatureFlux"/>')
 832         lines.append('    <var name="latentHeatFlux"/>')
 833         lines.append('    <var name="sensibleHeatFlux"/>')
 834         lines.append('    <var name="longWaveHeatFluxUp"/>')
 835         lines.append('    <var name="longWaveHeatFluxDown"/>')
 836         lines.append('    <var name="seaIceHeatFlux"/>')
 837         lines.append('    <var name="shortWaveHeatFlux"/>')
 838         lines.append('    <var name="evaporationFlux"/>')
 839         lines.append('    <var name="seaIceSalinityFlux"/>')
 840         lines.append('    <var name="seaIceFreshWaterFlux"/>')
 841         lines.append('    <var name="riverRunoffFlux"/>')
 842         lines.append('    <var name="iceRunoffFlux"/>')
 843         lines.append('    <var name="rainFlux"/>')
 844         lines.append('    <var name="snowFlux"/>')
 845         lines.append('    <var name="vertNonLocalFlux"/>')
 846         lines.append('    <var_array name="activeTracersTend"/>')
 847         lines.append('    <var name="salinitySurfaceRestoringTendency"/>')
 848         lines.append('    <var_array name="activeTracerHorizontalAdvectionTendency"/>')
 849         lines.append('    <var_array name="activeTracerVerticalAdvectionTendency"/>')
 850         lines.append('    <var_array name="activeTracerVertMixTendency"/>')
 851         lines.append('    <var_array name="activeTracerSurfaceFluxTendency"/>')
 852         lines.append('    <var_array name="temperatureShortWaveTendency"/>')
 853         lines.append('    <var_array name="activeTracerNonLocalTendency"/>')
 854         lines.append('    <var name="areaCellGlobal"/>')
 855         lines.append('    <var name="areaEdgeGlobal"/>')
 856         lines.append('    <var name="areaTriangleGlobal"/>')
 857         lines.append('    <var name="volumeCellGlobal"/>')
 858         lines.append('    <var name="volumeEdgeGlobal"/>')
 859         lines.append('    <var name="CFLNumberGlobal"/>')
 860         lines.append('    <var name="vertDiffTopOfCell"/>')
 861         lines.append('    <var name="vertViscTopOfCell"/>')
 862         lines.append('    <var name="bulkRichardsonNumber"/>')
 863         lines.append('    <var name="boundaryLayerDepth"/>')
 864         lines.append('    <var name="columnIntegratedSpeed"/>')
 865         lines.append('    <var name="landIceFreshwaterFlux"/>')
 866         lines.append('    <var name="landIceHeatFlux"/>')
 867         lines.append('    <var name="heatFluxToLandIce"/>')
 868         lines.append('    <var name="mocStreamvalLatAndDepth"/>')
 869         lines.append('    <var name="mocStreamvalLatAndDepthRegion"/>')
 870         lines.append('    <var name="binBoundaryMocStreamfunction"/>')
 871         lines.append('    <var name="surfaceBuoyancyForcing"/>')
 872         lines.append('    <var name="tendLayerThickness"/>')
 873         lines.append('    <var name="boundaryLayerDepthSmooth"/>')
 874         lines.append('    <var name="pressureAdjustedSSH"/>')
 875         lines.append('    <var name="SSHSquared"/>')
 876         lines.append('    <var name="velocityZonalSquared"/>')
 877         lines.append('    <var name="velocityMeridionalSquared"/>')
 878         lines.append('    <var name="velocityZonalTimesTemperature"/>')
 879         lines.append('    <var name="velocityMeridionalTimesTemperature"/>')
 880         lines.append('    <var_array name="activeTracerVerticalAdvectionTopFlux"/>')
 881         lines.append('    <var_array name="activeTracerHorizontalAdvectionEdgeFlux"/>')
 882         if ocn_iceberg == 'true':
 883             lines.append('    <var name="icebergHeatFlux"/>')
 884             lines.append('    <var name="icebergFreshWaterFlux"/>')

 886         if ocn_bgc in ['eco_only', 'eco_and_dms', 'eco_and_macromolecules', 'eco_and_dms_and_macromolecules']:
 887             lines.append('    <var name="PH_PREV_ALT_CO2_3D"/>')
 888             lines.append('    <var name="dust_FLUX_IN"/>')
 889             lines.append('    <var name="IRON_FLUX_IN"/>')
 890             lines.append('    <var name="PAR_surface"/>')
 891             lines.append('    <var name="windSpeedSquared10m"/>')
 892             lines.append('    <var name="atmosphericCO2"/>')
 893             lines.append('    <var name="atmosphericCO2_ALT_CO2"/>')
 894             lines.append('    <var name="CO2_gas_flux"/>')
 895             lines.append('    <var name="CO2_alt_gas_flux"/>')
 896             lines.append('    <var name="iceFluxPhytoC"/>')
 897             lines.append('    <var name="iceFluxNO3"/>')
 898             lines.append('    <var name="iceFluxSiO3"/>')
 899             lines.append('    <var name="iceFluxNH4"/>')
 900             lines.append('    <var name="iceFluxFeDissolved"/>')
 901             lines.append('    <var name="iceFluxDOC"/>')
 902             lines.append('    <var name="ecosys_diag_photoC_TOT_zint"/>')
 903             lines.append('    <var name="ecosys_diag_O2_ZMIN"/>')
 904             lines.append('    <var name="ecosys_diag_O2_ZMIN_DEPTH"/>')
 905             lines.append('    <var name="ecosys_diag_Jint_Ctot"/>')
 906             lines.append('    <var name="ecosys_diag_Jint_100m_Ctot"/>')
 907             lines.append('    <var name="ecosys_diag_Jint_Ntot"/>')
 908             lines.append('    <var name="ecosys_diag_Jint_100m_Ntot"/>')
 909             lines.append('    <var name="ecosys_diag_Jint_Ptot"/>')
 910             lines.append('    <var name="ecosys_diag_Jint_100m_Ptot"/>')
 911             lines.append('    <var name="ecosys_diag_Jint_Sitot"/>')
 912             lines.append('    <var name="ecosys_diag_Jint_100m_Sitot"/>')
 913             lines.append('    <var name="ecosys_diag_photoC_zint"/>')
 914             lines.append('    <var name="ecosys_diag_photoC_NO3_zint"/>')
 915             lines.append('    <var name="ecosys_diag_PAR_avg"/>')
 916             lines.append('    <var name="ecosys_diag_POC_FLUX_IN"/>')
 917             lines.append('    <var name="ecosys_diag_CaCO3_FLUX_IN"/>')
 918             lines.append('    <var name="ecosys_diag_auto_graze_TOT"/>')
 919             lines.append('    <var name="ecosys_diag_zoo_loss"/>')
 920             lines.append('    <var name="ecosys_diag_photoC_TOT"/>')
 921             lines.append('    <var name="ecosys_diag_photoC_NO3_TOT"/>')
 922             lines.append('    <var name="ecosys_diag_NITRIF"/>')
 923             lines.append('    <var name="ecosys_diag_DENITRIF"/>')
 924             lines.append('    <var name="ecosys_diag_calcToSed"/>')
 925             lines.append('    <var name="ecosys_diag_pocToSed"/>')
 926             lines.append('    <var name="ecosys_diag_SedDenitrif"/>')
 927             lines.append('    <var name="ecosys_diag_tot_Nfix"/>')
 928             lines.append('    <var name="ecosys_diag_O2_PRODUCTION"/>')
 929             lines.append('    <var name="ecosys_diag_O2_CONSUMPTION"/>')
 930             lines.append('    <var name="ecosys_diag_pH_3D"/>')
 931             lines.append('    <var name="ecosys_diag_POC_PROD"/>')
 932             lines.append('    <var name="ecosys_diag_POC_REMIN"/>')
 933             lines.append('    <var name="ecosys_diag_POC_ACCUM"/>')
 934             lines.append('    <var name="ecosys_diag_N_lim"/>')
 935             lines.append('    <var name="ecosys_diag_P_lim"/>')
 936             lines.append('    <var name="ecosys_diag_Fe_lim"/>')
 937             lines.append('    <var name="ecosys_diag_SiO3_lim"/>')
 938             lines.append('    <var name="ecosys_diag_light_lim"/>')
 939             lines.append('    <var name="ecosys_diag_photoC"/>')
 940             lines.append('    <var name="ecosys_diag_photoC_NO3"/>')
 941             lines.append('    <var name="ecosys_diag_photoFe"/>')
 942             lines.append('    <var name="ecosys_diag_photoNO3"/>')
 943             lines.append('    <var name="ecosys_diag_photoNH4"/>')
 944             lines.append('    <var name="ecosys_diag_DOP_uptake"/>')
 945             lines.append('    <var name="ecosys_diag_PO4_uptake"/>')
 946             lines.append('    <var name="ecosys_diag_auto_graze"/>')
 947             lines.append('    <var name="ecosys_diag_auto_loss"/>')
 948             lines.append('    <var name="ecosys_diag_auto_agg"/>')
 949             lines.append('    <var name="ecosys_diag_CO2star"/>')
 950             lines.append('    <var name="ecosys_diag_dCO2star"/>')
 951             lines.append('    <var name="ecosys_diag_pCO2surface"/>')
 952             lines.append('    <var name="ecosys_diag_dpCO2"/>')
 953             lines.append('    <var name="ecosys_diag_pistonVel_CO2"/>')
 954             lines.append('    <var name="ecosys_diag_Schmidt_CO2"/>')
 955             lines.append('    <var name="ecosys_diag_xkw"/>')
 956             lines.append('    <var name="ecosys_diag_pistonVel_O2"/>')
 957             lines.append('    <var name="ecosys_diag_O2_saturation"/>')
 958             lines.append('    <var name="ecosys_diag_Schmidt_O2"/>')
 958             lines.append('    <var name="ecosys_diag_CO2star_ALT_CO2"/>')
 959             lines.append('    <var name="ecosys_diag_dCO2star_ALT_CO2"/>')
 960             lines.append('    <var name="ecosys_diag_pCO2surface_ALT_CO2"/>')
 961             lines.append('    <var name="ecosys_diag_dpCO2_ALT_CO2"/>')
 962             lines.append('    <var name="ecosys_diag_tot_bSi_form"/>')
 963             lines.append('    <var name="ecosys_diag_DOC_prod"/>')
 964             lines.append('    <var name="ecosys_diag_DOC_remin"/>')
 965             lines.append('    <var name="ecosys_diag_CaCO3_form_zint"/>')
 966             lines.append('    <var name="ecosys_diag_CaCO3_PROD"/>')
 967             lines.append('    <var name="ecosys_diag_CaCO3_REMIN"/>')
 968             lines.append('    <var name="ecosys_diag_tot_CaCO3_form"/>')
 969             lines.append('    <var name="ecosys_diag_co3_sat_calc"/>')
 970             lines.append('    <var name="ecosys_diag_co3_sat_arag"/>')
```
</details>

mark-petersen commented:
@milenaveneziani and @xylar should we add the bins (MOC, MHT) to the time average file? Anything else for more convenience in post-processing?

xylar commented:
@mark-petersen, yes the binds would be good to include.  One less set of files to worry about in MPAS-Analysis if it's in timeSeriesStatsMonthly.

Regarding `zMid`, we had discussed before that it's a large 3D field so many felt that it was too much space to justify for something that can be post-processed.  I disagreed at the time but conceded the point.  I just want to make sure those who felt that way now think it's justified (or negligible compared to other output).

rljacob commented:
@jonbob don't forget about this.

jonbob commented:
@rljacob - it's on my list. But this one is waiting for several other people to review first.

maltrud commented:
@mark-petersen i thought we had talked about adding atmosphericPressure, but i don't see it?  this would be for both monthly and high frequency output.

rljacob commented:
@maltrud @vanroekel @milenaveneziani please review.

milenaveneziani commented:
I seem to understand that `zMid` was added already. I think that should be fine.

Also, @vanroekel is adding mixed layer budget terms as online-calculated quantities: I would suggest to add those terms to the monthly fields (they are all 2-d quantities).

maltrud commented:
as i noted earlier, i would like atmosphericPressure added to both monthly and high frequency.

rljacob commented:
@milenaveneziani :  @xylar was saying that zMid isn't needed.  Are you saying that it is?

@jonbob please add atmosphericPressure as @maltrud requested.

vanroekel commented:
I would like to make sure we have the mixed layer budgets mentioned by @milenaveneziani also I would like to add BruntVaisalaFreqTop to the monthly stream

rljacob commented:
From IG telecon:  still figuring out what is needed here.

rljacob commented:
@vanroekel what about this PR and the upcoming v2 sims?

vanroekel reviewed (changes requested):
@mark-petersen I think we need a few more fields, here's what I'd like, many of these are on master now, so maybe a rebase of this branch would fix things
```
<var_array name="activeTracerHorMixTendency"/>
     <var name="oceanHeatContentSfcToBot"/>
     <var name="oceanHeatContentSfcTo700m"/>
     <var name="oceanHeatContent700mTo2000m"/>
     <var name="oceanHeatContent2000mToBot"/>
     <var name="activeTracerForcingMLTend"/>
     <var name="activeTracerHorMixMLTend"/>
     <var name="activeTracerNonLocalMLTend"/>
     <var name="activeTracerVertAdvectionMLTend"/>
     <var name="activeTracerHorAdvectionMLTend"/>
     <var name="activeTracerVertMixMLTend"/>
     <var name="activeTracersML"/>
     <var name="bruntVaisalaFreqML"/>
     <var name="activeTracersTendML"/>
<var name="BruntVaisalaFreqTop"/>
```

mark-petersen commented:
@vanroekel I rebased and checked through the comments above. Most were already added, as you said. I added `zMid`, `atmosphericPressure` per @maltrud's request, the missing bins variable. The only one on your list to add to the monthly averages was `BruntVaisalaFreqTop`.

@maltrud and @milenaveneziani can you also double check and approve?

vanroekel reviewed (approved):
looks good, thanks @mark-petersen

mark-petersen commented:
Ran a test to check for typos. Passes
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_intel
```

milenaveneziani reviewed (approved):
looks good to me

xylar reviewed (approved):
I'm fine with these additions.

mark-petersen commented:
Current streams file from PR head using a G case:
[streams.ocean_200924a.txt](https://github.com/E3SM-Project/E3SM/files/5277969/streams.ocean_200924a.txt)

maltrud commented:
i have 2 possibilities to add.  1) kappa for GM if we're using a variable GM.  but we don't know if that's what we'll be using.  we could use the run script to add or subtract it.  or just output it even if it is constant.  2) is there a Redi tendency array that we can output?

mark-petersen commented:
(1) These are all the arrays used for GM and Redi. I would need to think about what would be useful in output. @maltrud do you use any of these now?
```
grep -i  kappa src/core_ocean/Registry.xml
		<var name="gmKappaScaling" type="real" dimensions="nVertLevelsP1 nCells Time" units="NA"
			description="spatially and depth varying GM kappa.  The scaling is based on the Brunt Vaisala Frequency relative to a maximum value below the mixed layer, follows from Danabasoglu and Marshall 2007.  If config_GM_closure is not set to N2_dependent the scaling value is set to 1 everywhere."
		<var name="RediKappaScaling" type="real" dimensions="nVertLevelsP1 nCells Time" units="NA"
			description="Scaling coefficient for GM kappa. Varies from 0 to 1.  The scaling is based on the Brunt Vaisala Frequency relative to a maximum value below the mixed layer, follows from Danabasoglu and Marshall 2007.  If config_Redi_use_N2_based_taper = .false. the scaling is set to 1 everywhere."
		<var name="RediKappaSfcTaper" type="real" dimensions="nVertLevelsP1 nCells Time" units="NA"
			description="Scaling coefficient for Redi kappa. Varies from 0 to 1."
		<var name="gmBolusKappa" type="real" dimensions="nEdges Time" units="m^2 s^{-1}"
			 description="GM Bolus Kappa value.  This is constant in time, and set at init based on the constant or ramp fuction."
		<var name="RediKappa" type="real" dimensions="nEdges Time" units="m^2 s^{-1}"
			 description="Redi Kappa value.  This is constant in time, and set at init based on the constant or ramp fuction."
```

(2) We don't have an array for Redi tendency by itself. That is included in the horizontal and vertical tendency:
```
grep Tend src/core_ocean/Registry.xml |grep -i var|grep Mix
<var_array name="activeTracerHorMixTendency" type="real" dimensions="nVertLevels nCells Time" packages="tracerBudget">
  <var name="temperatureHorMixTendency" array_group="activeTracerHmixTendGroup" units="degrees Celsius per second" name_in_code="temperatureHorMixTendency"
  <var name="salinityHorMixTendency" array_group="activeTracerHmixTendGroup" units="PSU per second" name_in_code="salinityHorMixTendency"
<var_array name="activeTracerVertMixTendency" type="real" dimensions="nVertLevels nCells Time" packages="tracerBudget">
  <var name="temperatureVertMixTendency" array_group="activeTracerVertMixTendGroup" units="degrees Celsius per second" name_in_code="temperatureVertMixTendency"
  <var name="salinityVertMixTendency" array_group="activeTracerVertMixTendGroup" units="PSU per second" name_in_code="salinityVertMixTendency"
```

vanroekel commented:
The only ones I would suggest adding are ```gmKappaScaling``` and ```gmBolusKappa```

mark-petersen commented:
OK, I added those gm variables. @maltrud if you are happy with these variables, please approve. @jonbob this can be merged after that.

mark-petersen commented:
@jonbob this can now be merged when convenient.

golaz commented:
@rljacob : this is needed to finalize v2 WaterCycle and should go on master when ready.

jonbob commented:
merged to next

jonbob commented:
merged to master
